aaelf64/aaelf64.rst

```diff
@@ -1249,6 +1249,10 @@ Finally the bit-field of X specified in the table (those bits of X picked out by
 
 For "``MOVW``" type relocations it is the assembler’s responsibility to encode the hw bits (bits 21 and 22) to indicate the bits in the target value that the immediate field represents.
 
+Relocation relaxation
+^^^^^^^^^^^^^^^^^^^^^
+
+Linkers may perform relocation relaxations to optimize instructions affected by relocation. Relocations that result in addition with zero may emit a faster register move or nop instead. The relocations ``R_<CLS>_ADR_GOT_PAGE`` and ``R_<CLS>_LD64_GOT_LO12_NC`` may be relaxed if all such relocations for a symbol are consecutive pairs, and both relocations write the same register. In this case each pair is independent and can be safely relaxed. If the symbol is not pre-emptable, the addend is zero and the destination is within ADRP range, the relocations relax to ``R_<CLS>_ADR_PREL_PG_HI21`` and ``R_<CLS>_LDST64_ABS_LO12_NC``.
```

Comment:
I think it would be useful to split this up and make some of the conditions more formal. I've made a suggestion below.

Linkers may optionally perform relocation relaxations to optimize instructions affected by relocation. This can take one of two forms:
* Relocations that operate on ADD (immediate) instructions where the result of the Operation is 0 may replace the ADD (immediate) instruction with a MOV (register) or NOP.
* A sequence of ordered instructions and relocations can be relaxed to an alternative sequence of ordered instructions and relocations when all of the following are true:
  * The instructions in the sequence are consecutive and in the correct order.
  * Each instruction has a relocation directive.
  * The relocation directives are in the same order as the instructions.
  * The conditions of the relaxation are met.
  
The ABI identifies the following relaxation sequences:

Relaxation of GOT-Relative to PC-Relative addressing mode.
Instructions and Relocations to be relaxed:
```
ADRP rd, :got: symbol                // R_<CLS>_ADR_GOT_PAGE 
LD      rd, [rd :got_lo12: symbol]   // R_<CLS>_LD64_GOT_LO12_NC
```
Relaxed instruction sequence:
```
ADRP rd, symbol                  // R_<CLS>_ADR_PREL_PG_HI21 
LD      rd, [rd :lo12: symbol]   // R_<CLS>_LD64_ABS_LO12_NC
```
Conditions for the relaxation:
* ``rd`` is the same register.
* ``symbol`` is a non-preemptible definition.
* The ``symbol`` does not have type ``STT_GNU_IFUNC``.
* The relocation addend is 0.
* The result of the relaxed relocations are within range.

Reply:
Other thoughts:
* I've suggested not doing the relaxation for ifuncs resolvers as this causes more problems (canonical PLT). This could be reduced to a recommendation or omitted as QoI but thought it simple enough to just say don't do it.
* (Not for this patch) The TLS relaxations should ideally be documented here, or the sysvabi64 (by reference).

One possible objection to applying the relaxation is the intent of the compiler/assembler. For example did they use a GOT generating relocation because they wanted to force the generation of a GOT entry, or did they just want to load the address of a variable via the most efficient means necessary. I can't currently think of a reason to require a GOT entry when PC-relative addressing would suffice but it may exist.

Another possibility that we could do instead of, or even in addition to is to introduce a new set of relocation that explicitly permit relaxation. These would obviously be harder to deploy as it would require a new linker to work at all, not just to be optimised. I had originally thought it would be possible to define these new relocations such that it would be possible to avoid the constraint that the ADRP and LD were consecutive, however if they weren't then there is a risk that the ADRP and LD are in different 4k pages and one being in range and the other not.

Reply:
That's a good suggestion, I'll update the text. Note the pairs are necessary since both relocations must be relaxed in lockstep (you obviously can't relax one but not the other!). Having pairs also means the ADRP value of the first instruction cannot escape and be combined with a 2nd instruction elsewhere, ie. each pair is independent. Any new relocations would still need to define pairs - the only advantage of new relocations is that the linker could assume that all preconditions are met, which might save some checks.